docs/source/notes/extending.rst

```diff
@@ -484,6 +484,54 @@ a case the rules are:
 * If all of the ``__torch_function__`` implementations return
   ``NotImplemented``, PyTorch raises a ``TypeError``.
 
+Testing Coverage of Overrides for the PyTorch API
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+One troublesome aspect of implementing ``__torch_function__`` is that if some
+operations do and others do not have overrides, users will at best see an
+inconsistent experience, or at worst will see errors raised at runtime when they
+use a function that does not have an override. To ease this process, PyTorch
+provides a developer-facing API for ensuring full support for
+``__torch_function__`` overrides. This API is private and may be subject to
+changes without warning in the future.
+
+First, to get a listing of all overridable functions, use
+``torch._overrides.get_overridable_functions``. This returns a dictionary whose
+keys are namespaces in the ``PyTorch`` Python API and whose values are a list of
+functions in that namespace that can be overrided. For example, let's print the
+names of the first 5 functions in ``torch.nn.functional`` that can be
+overrided::
+
+  >>> from torch._overrides import get_overridable_functions
+  >>> func_dict = get_overridable_functions()
+  >>> nn_funcs = func_dict[torch.nn.functional]
+  >>> print([f.__name__ for f in nn_funcs[:5])
+  ['adaptive_avg_pool1d', 'adaptive_avg_pool2d', 'adaptive_avg_pool3d',
+   'adaptive_max_pool1d', 'adaptive_max_pool1d_with_indices']
+
+This listing of functions makes it possible to iterate over all overridable
+functions, however in practice this is not enough to write tests for all of
+these functions without laboriously and manually copying the signature of each
+function for each test. To ease this process, the
+``torch._overrides.get_testing_overrides`` function returns a dictionary mapping
```

Comment:
I see in the test below you exercise the torch.mean function, but there are two overloads for torch.mean

```
torch.mean(input) → Tensor
```

and

```
torch.mean(input, dim, keepdim=False, out=None) → Tensor
```

How is the second one tests?

Further we have our [own argparser](https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/python_arg_parser.cpp#L594), which means the Python signature of a given function might not accurately capture all capabilities.

How is this dealt with?

Reply:
`mean` is a C++ kernel, so its handled via the C++ argument parsing, e.g. https://github.com/pytorch/pytorch/blob/b10761d8900af1dc48bee807ab39c80fd4334b79/torch/csrc/utils/python_arg_parser.cpp#L247-L251.

I don't think there's a direct test for both signatures in the override tests. In principle we could add it but I'm not sure if it would be worth the extra boilerplate.